harmonizer/src/lib.rs

```diff
@@ -0,0 +1,216 @@
+use deno_core::Op;
+use deno_core::{json_op_sync, JsRuntime};
+use serde::{Deserialize, Serialize};
+use std::sync::mpsc::channel;
+use std::{fmt::Display, io::Write};
+use thiserror::Error;
+
+#[derive(Serialize)]
+// When serialized, we'll be putting this into JavaScript expecting camelCase.
+#[serde(rename_all = "camelCase")]
+pub struct ServiceDefinition {
+    pub name: String,
+    pub url: String,
+    pub type_defs: String,
+}
+
+impl ServiceDefinition {
+    pub fn new<N: Into<String>, U: Into<String>, D: Into<String>>(
+        name: N,
+        url: U,
+        type_defs: D,
+    ) -> ServiceDefinition {
+        ServiceDefinition {
+            name: name.into(),
+            url: url.into(),
+            type_defs: type_defs.into(),
+        }
+    }
+}
+
+pub type ServiceList = Vec<ServiceDefinition>;
+
+#[derive(Debug, Error, Serialize, Deserialize, PartialEq)]
+pub struct CompositionError {
+    pub message: Option<String>,
+    pub extensions: Option<CompositionErrorExtensions>,
+}
+
+impl Display for CompositionError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if let Some(msg) = &self.message {
+            f.write_fmt(format_args!("{code}: {msg}", code = self.code(), msg = msg))
+        } else {
+            f.write_str(self.code())
+        }
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct CompositionErrorExtensions {
+    pub code: String,
+}
+
+impl CompositionError {
+    pub fn code(&self) -> &str {
+        if let Some(ref ext) = self.extensions {
+            ext.code.as_str()
```

Comment:
if you like, you can also do: 

```suggestion
        if let Some(ref ext) = self.extensions {
            &*ext.code
```

Reply:
Fixed in ea0c9a0d7deccb150fb4966a4487de1a10dd444e, which also switched this to `match`.